### ISSUE TITLE
C++ Can only index with tensors that are scalars (zero-dim)

### ISSUE THREAD

**Issue by Terodee** [pytorch/pytorch#20263]
I got an error ```Can only index with tensors that are scalars (zero-dim) (operator [] at C:\w\1\s\windows\pytorch\aten\src\ATen/TensorOperators.h:62)(no backtrace available)```
while the python one ran normal:
```python
torch.Size([2, 2, 1])
tensor([[[[[1],
           [1]]],

         [[[1],
           [1]]]],

        [[[[1],
           [1]]],

         [[[1],
           [1]]]]])
```
code is shown below
```c++
#include <torch/torch.h>
#include <iostream>
using namespace std;
int main()
{
	try
	{
		torch::Tensor data = torch::eye(3);
		torch::Tensor index = torch::eye(3);
		data[index];
	}
	catch (const std::exception&e)
	{
		cout << e.what();
	}
	getchar();
	return 0;
}
```

```python 
import torch
import torch.nn as nn
data=torch.tensor([[[1],[1]],[[1],[1]]]);
index=torch.tensor([[[1],[1]],[[1],[1]]]);
print(data.shape)
print(data[index])
```

is there anything i didn't notice about the C++ frontend?or the interface is a little different from the python one.

**Comment by Terodee** [comment]
![08Q)C@5X{EAHB_AYQN}@KAW](https://user-images.githubusercontent.com/12407152/57351503-b195e500-7194-11e9-9d73-233d7c3fb37d.png)

**Comment by yf225** [comment]
Tensors in C++ API currently only support indexing with scalar / zero-dim tensor. We have plans to support tensor slicing, advanced indexing and `tensor[mask]` syntax in the future.

**Comment by sto-chastic** [comment]
any recommended workaround? My code looks like this:
`    auto depth_slices = torch::zeros({num_maps, height, width}, at::kInt)
        .to(this->device);

    int prev_depth = 0;
    for (int i = 0; i < num_maps; i++) {
        float curr_depth = depth_range[i].item<float>();
        auto depth_mask = 1 - depth.gt(curr_depth);
        depth_mask = depth_mask * depth.gt(prev_depth);

        depth_slices[i][depth_mask] = imageL[depth_mask].to(at::kInt);
    }`
where I need to select the indexes that satisfy some conditions and store the respective values on a second tensor in the same index location

**Comment by jackeown** [comment]
This would be a very nice feature to have.

**Comment by yf225** [comment]
This feature is available in libtorch v1.5. Following the docs in https://pytorch.org/cppdocs/notes/tensor_indexing.html, we can write
```python
data=torch.tensor([[[1],[1]],[[1],[1]]]);
index=torch.tensor([[[1],[1]],[[1],[1]]]);
print(data[index])
```
in C++ as
```cpp
auto data = torch::tensor({{{1}, {1}}, {{1}, {1}}});
auto index = torch::tensor({{{1}, {1}}, {{1}, {1}}});
std::cout << data.index({index}) <<std::endl;
```

**Comment by jackeown** [comment]
Do you have to do `data.index({index})` instead of `data[index]`?

**Comment by mmbejani** [comment]
`data.index` is dramatically slow (in my experiments about 10x!!!!!!!)
